lib/assets/chart_builder/main.js

```diff
@@ -0,0 +1,223 @@
+  setValues(info.fields);
+
+  const hasType = ["x_axis", "y_axis", "color"];
+  const types = ["x_axis_type", "y_axis_type", "color_type"];
+
+  ctx.root.addEventListener("blur", handleFieldChange, true);
+  ctx.root.addEventListener("change", handleFieldChange);
+
+  function handleFieldChange(event) {
+    const { name, value } = event.target;
+    if (hasType.includes(name)) { maybeReset({name, value}) }
+    ctx.pushEvent("update_field", { field: name, value });
+  }
+
+  ctx.handleEvent("update", ({ fields }) => {
+    setValues(fields);
+  });
+
+  ctx.handleEvent("missing_dep", ({ dep }) => {
+    updateInfoBox(dep);
+  });
+
+  ctx.handleEvent("set_axis_options", ({ options }) => {
+    options ? setAxisOptions(options) : missingData();
+  });
+
+  ctx.handleEvent("set_data_options", ({ options }) => {
+    setDataOptions(options);
+  });
+
+  function setValues(fields) {
+    for (const field in fields) {
+      ctx.root.querySelector(`[name="${field}"]`).value = fields[field];
+    }
+  }
+
+  function setDataOptions(options) {
+    const selectElem = ctx.root.querySelector(`[name="data"]`);
+    selectElem.innerHTML = "";
+    createOptions(selectElem, Object.keys(options))
+    const selected = selectElem.value;
+    updateDataInfoBox(options);
+    ctx.pushEvent("update_field", { field: "data", value: selected });
+  }
+
+  function setAxisOptions(options) {
+    for (const elem of hasType) {
+      const selectElem = ctx.root.querySelector(`[name="${elem}"]`);
+      selectElem.disabled = false;
+      selectElem.innerHTML = "<option value=''></option>";
+      createOptions(selectElem, options)
+      elem === "color" ? selectElem.selectedIndex = 0 : selectElem.selectedIndex = 1;
+      const selected = selectElem.value;
+      ctx.pushEvent("update_field", { field: elem, value: selected });
+    }
+    setTypes();
+  }
+
+  function setTypes() {
+    for (const elem of types) {
+      const selectElem = ctx.root.querySelector(`[name="${elem}"]`);
+      selectElem.selectedIndex = 0;
+      const selected = selectElem.value;
+      elem === "color_type" ? selectElem.disabled = true : selectElem.disabled = false;
+      ctx.pushEvent("update_field", { field: elem, value: selected });
+    }
+  }
+
+  function maybeReset( {name, value} ) {
+    const elem = ctx.root.querySelector(`[name="${name}_type"]`)
+    value === "" ? resetField(elem) : elem.disabled = false;
+  }
+
+  function missingData() {
+    for (const name of hasType.concat(types)) {
+      const elem = ctx.root.querySelector(`[name="${name}"]`);
+      resetField(elem);
+    }
+    updateDataInfoBox();
+  }
+
+  function resetField(elem) {
+    elem.value = "";
+    elem.disabled = true;
+    ctx.pushEvent("update_field", { field: elem.name, value: "" });
+  }
+
+  function createOptions(selectElem, options) {
+    for (const option of options) {
+      const element = document.createElement("option");
+      element.innerText = option;
```

Comment:
This is cool but I need more help here! I ended up with more code for the elements where I need to append or prepend options. Or with two different ways to create the options depending on the rules.

Reply:
I'd say it's enough to make the select dimmed, rather than individual options (that's what we do for the SQL cell). Then you could conditionally prepend an option to the list and render all of them. That said, I'm ok with either approach, and creating elements directly is fine, so your call :)

Reply:
I'll keep it as it is for now! :)